### ISSUE TITLE
Data Leakage and Overfitting Issue

### ISSUE THREAD

**Issue by a1473838623** [shaieesss/TransDe#2]
### Issues Found:

1. In `solver.py` line 142:
   ```python
   vali_loss1, vali_loss2 = self.vali(self.test_loader)
   ```
   This causes test data leakage and overfitting problems as it directly uses the test set for validation.

2. The `get_loader_segment` function in `data_loader.py` does not properly separate a validation set from the training set. Instead, it uses the test set as the validation set, which leads to data leakage.

### Suggested Fix:

1. Modify `solver.py` line 142 to use a proper validation loader:
   ```python
   vali_loss1, vali_loss2 = self.vali(self.vali_loader) # change from self.test_loader
   ```

2. Update the `get_loader_segment` function in `data_loader.py` to properly split training data into training and validation sets:
   ```python
   def get_loader_segment(index, data_path, batch_size, win_size=100, step=100, mode='train', dataset='KDD', val_ratio=0.2):
       ori_mode = mode
       if mode == 'train' or mode == 'val':
           mode = 'train'
       if (dataset == 'SMD'):
           dataset = SMDSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'MSL'):
           dataset = MSLSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'SMAP'):
           dataset = SMAPSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'PSM'):
           dataset = PSMSegLoader(data_path, win_size, 1, mode)
       elif (dataset =='SWAT'):
           dataset = SWATSegLoader(data_path,win_size,1,mode)
       elif (dataset == 'UCR'):
           dataset = UCRSegLoader(index, data_path, win_size, 1, mode)
       elif (dataset == 'UCR_AUG'):
           dataset = UCRAUGSegLoader(index, data_path, win_size, 1, mode)
       elif (dataset == 'NIPS_TS_Water'):
           dataset = NIPS_TS_WaterSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'NIPS_TS_Swan'):
           dataset = NIPS_TS_SwanSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'NIPS_TS_CCard'):
           dataset = NIPS_TS_CCardSegLoader(data_path, win_size, 1, mode)
       elif (dataset == 'SMD_Ori'):
           dataset = SMD_OriSegLoader(index, data_path, win_size, 1, mode)
       if ori_mode == 'train' or ori_mode == 'val':
           # Separate validation set from training data
           full_size = len(dataset)
           train_size = int(full_size * (1 - val_ratio))
           val_size = full_size - train_size
           
           # Use random_split to divide the dataset
           train_dataset, val_dataset = torch.utils.data.random_split(
               dataset, [train_size, val_size]
           )
           if ori_mode== 'train':
               return DataLoader(train_dataset, batch_size=batch_size, shuffle=True, num_workers=8, drop_last=False)
           else:  # ori_mode== 'val'
               return DataLoader(val_dataset, batch_size=batch_size, shuffle=True, num_workers=8, drop_last=False)
       else:  # ori_mode== 'test'
           return DataLoader(dataset, batch_size=batch_size, shuffle=False, num_workers=8, drop_last=False)
   ```
3. Modify `solver.py` line 176 to use a proper validation loader:
   ```python
   for i, (input_data, labels) in enumerate(self.vali_loader): # change from self.thre_loader
   ```
This will prevent data leakage between test and validation sets, ensuring a more reliable model evaluation process.

**Comment by shaieesss** [comment]
Thanks for the help. I will revise soon，because recently I am little busy with some works.

| |
msezwx
|
|
***@***.***
|
---- Replied Message ----
| From | ***@***.***> |
| Date | 05/10/2025 10:55 |
| To | shaieesss/TransDe ***@***.***> |
| Cc | Subscribed ***@***.***> |
| Subject | [shaieesss/TransDe] Data Leakage and Overfitting Issue (Issue #2) |
a1473838623 created an issue (shaieesss/TransDe#2)
Issues Found:

In solver.py line 142:

vali_loss1, vali_loss2=self.vali(self.test_loader)

This causes test data leakage and overfitting problems as it directly uses the test set for validation.

The get_loader_segment function in data_loader.py does not properly separate a validation set from the training set. Instead, it uses the test set as the validation set, which leads to data leakage.

Suggested Fix:

Modify solver.py line 142 to use a proper validation loader:

vali_loss1, vali_loss2=self.vali(self.vali_loader) # change from self.test_loader

Update the get_loader_segment function in data_loader.py to properly split training data into training and validation sets:

defget_loader_segment(index, data_path, batch_size, win_size=100, step=100, mode='train', dataset='KDD', val_ratio=0.2):
    ori_mode=modeifmode=='train'ormode=='val':
        mode='train'if (dataset=='SMD'):
        dataset=SMDSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='MSL'):
        dataset=MSLSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='SMAP'):
        dataset=SMAPSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='PSM'):
        dataset=PSMSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='SWAT'):
        dataset=SWATSegLoader(data_path,win_size,1,mode)
    elif (dataset=='UCR'):
        dataset=UCRSegLoader(index, data_path, win_size, 1, mode)
    elif (dataset=='UCR_AUG'):
        dataset=UCRAUGSegLoader(index, data_path, win_size, 1, mode)
    elif (dataset=='NIPS_TS_Water'):
        dataset=NIPS_TS_WaterSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='NIPS_TS_Swan'):
        dataset=NIPS_TS_SwanSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='NIPS_TS_CCard'):
        dataset=NIPS_TS_CCardSegLoader(data_path, win_size, 1, mode)
    elif (dataset=='SMD_Ori'):
        dataset=SMD_OriSegLoader(index, data_path, win_size, 1, mode)
    ifori_mode=='train'orori_mode=='val':
        # Separate validation set from training datafull_size=len(dataset)
        train_size=int(full_size* (1-val_ratio))
        val_size=full_size-train_size# Use random_split to divide the datasettrain_dataset, val_dataset=torch.utils.data.random_split(
            dataset, [train_size, val_size]
        )
        ifori_mode=='train':
            returnDataLoader(train_dataset, batch_size=batch_size, shuffle=True, num_workers=8, drop_last=False)
        else:  # ori_mode== 'val'returnDataLoader(val_dataset, batch_size=batch_size, shuffle=True, num_workers=8, drop_last=False)
    else:  # ori_mode== 'test'returnDataLoader(dataset, batch_size=batch_size, shuffle=False, num_workers=8, drop_last=False)

Modify solver.py line 176 to use a proper validation loader:

fori, (input_data, labels) inenumerate(self.vali_loader): # change from self.thre_loader

This will prevent data leakage between test and validation sets, ensuring a more reliable model evaluation process.
